blockchain_connector/avalon_blockchain_connector/ethereum/ethereum_connector.py

```diff
@@ -98,48 +98,94 @@ def _retrieve_first_worker_details(self):
+    def _add_update_worker_to_chain(self, wids_onchain, wids_kv,
+                                    jrpc_worker_registry):
         """
-        This function adds a worker to the Ethereum blockchain
+        This function adds/updates a worker in the Ethereum blockchain
         """
-        worker_id = self._eth_client.get_bytes_from_hex(worker_id)
-        worker_type = worker_info["workerType"]
-        org_id = self._eth_client\
-            .get_bytes_from_hex(worker_info["organizationId"])
-        app_type_id = self._eth_client.get_bytes_from_hex(
-            worker_info["applicationTypeId"])
-        details = json.dumps(worker_info["details"])
-
-        txn_dict = self._worker_reg_contract_instance.functions.workerRegister(
-            worker_id, worker_type, org_id, [app_type_id], details)\
-            .buildTransaction(self._eth_client.get_transaction_params())
-        try:
-            txn_receipt = self._eth_client.execute_transaction(txn_dict)
-        except Exception as e:
-            logging.error("Error while adding worker to ethereum"
-                          + " blockchain : "+str(e))
+
+        new_workers_kv = []
+        for wid in wids_kv:
+            worker_info = self._retrieve_worker_details_from_kv_storage(
+                jrpc_worker_registry, wid)
+            worker_id = self._eth_client.get_bytes_from_hex(wid)
+            worker_type = worker_info["workerType"]
+            org_id = self._eth_client\
+                .get_bytes_from_hex(worker_info["organizationId"])
+            app_type_id = self._eth_client.get_bytes_from_hex(
+                worker_info["applicationTypeId"])
+            details = json.dumps(worker_info["details"])
+
+            if wid in wids_onchain:
+                logging.info("Updating worker {} on ethereum blockchain"\
+                    .format(wid))
+                txn_dict = self._worker_reg_contract_instance.functions\
+                    .workerUpdate(worker_id, details)\
+                    .buildTransaction(self._eth_client.get_transaction_params())
+            else:
+                logging.info("Adding new worker {} to ethereum blockchain"\
+                    .format(wid))
+                txn_dict = self._worker_reg_contract_instance.functions\
+                    .workerRegister(worker_id, worker_type, org_id,
+                                    [app_type_id], details)\
+                    .buildTransaction(self._eth_client.get_transaction_params())
+                new_workers_kv.append(wid)
+            try:
+                txn_receipt = self._eth_client.execute_transaction(txn_dict)
+            except Exception as e:
+                logging.error("Error while adding/updating worker to ethereum"
+                              + " blockchain : "+str(e))
+        for wid in wids_onchain:
+            # Mark all stale workers on blockchain as decommissioned
+            if wid not in new_workers_kv:
```

Comment:
if a worker id is present in shared kv and on-chain then onchain worker gets updated.  All active workers present in shared KV and on-chain are valid. Only those workers which are not present in shared KV can be made stale. Looks like above logic only considers new workers in shared KV.